### ISSUE TITLE
ssh_exchange_identification: Connection closed by remote host

### ISSUE THREAD

**Issue by dafeda** [equinor/ert#2546]
**Describe the bug**

From Yammer:

ssh_exchange_identification: Connection closed by remote host

** Warning: could not find status of ******** using 'bjobs' - trying with 'bhist'.

Comment from user that might be of interest: 

"Suddenly the ert gui is not stable anymore.  any particulary reason? does not update the runtime either.  The jobs are running and i just have to keep track in the grid portal , but what has changed?"

Edit: 

User ran echo `$KOMODO_RELEASE` and got:

2021.12.00-py36

**Comment by dafeda** [comment]
Got another one on Yammer regarding this:

"I have earlier run 100 realizations without any problems but now I want to run 200 but after 119 realizations I get the message -> ssh_exchange_idetification: Connection closed by remote host. "

**Comment by markusdregi** [comment]
It would be interesting to have the ERT/komodo version being used as listed in the bug report template on this one. Would you be able to obtain that information?

**Comment by markusdregi** [comment]
The warnings stems from the lsf driver implementation and how it retrieves the status of the different jobs. There where some improvements done to the robustness of this code in the spring and a test case mocking lsf server failure was introduced: https://github.com/equinor/ert/tree/main/tests/ert_tests/lsf_queue

Due to this it would be very interesting to know which version they are experiencing these problems with... Also, you can use the test data linked to to reproduce a similar error consistently on your own machine :)

**Comment by dafeda** [comment]
User ran echo `$KOMODO_RELEASE` and got:

2021.12.00-py36

**Comment by markusdregi** [comment]
Good 👍🏻 

This typically happens when the lsf master nodes are unavailable due to overload or downtime. You can emulate the scenario using the LSF test referred to above. The GUI should not freeze on this, but it could be good to just to a test to verify...

Let me know if you need assistance.

**Comment by markusdregi** [comment]
Logs can be found [here](https://web.yammer.com/main/threads/eyJfdHlwZSI6IlRocmVhZCIsImlkIjoiMTUxMjU2NzkzMDU1MjMyMCJ9).

**Comment by BjarneHerland** [comment]
Summary of digging into the [provided logs from Jan 3](https://web.yammer.com/main/threads/eyJfdHlwZSI6IlRocmVhZCIsImlkIjoiMTUxMjU2NzkzMDU1MjMyMCJ9):

### Why does ERT abort?

Observe the reason spelled out in lines 33901+ in `ert-log 4.txt`

```
#2    Object "/prog/res/komodo/2021.12.01-py36-rhel7/root/lib64/python3.6/site-packages/res/.libs/libres.so", at 0x7fe590c2dd32, in enkf_state_complete_forward_model_EXIT_handler__
#1    Object "/prog/res/komodo/2021.12.01-py36-rhel7/root/lib64/python3.6/site-packages/res/.libs/libres.so", at 0x7fe590c546ea, in state_map_iset
#0    Object "/prog/res/komodo/2021.12.01-py36-rhel7/root/lib64/python3.6/site-packages/ecl/.libs/libecl.so", at 0x7fe5912f0cae, in util_abort__

with message: state_map_iset__: illegal state transition for realisation:32740 1 -> 8

from file: /github/workspace/libres/lib/enkf/state_map.cpp in state_map_iset__ at line: 150
```

This shows that ERT attempts transition from state 1 (`STATE_UNDEFINED`) to state 8 (`STATE_LOAD_FAILURE`) which is not allowed according to

https://github.com/equinor/ert/blob/5770c29261ed2d069220b509fb87916a60b136cb/libres/lib/enkf/state_map.cpp#L118-L119

triggering abort

https://github.com/equinor/ert/blob/5770c29261ed2d069220b509fb87916a60b136cb/libres/lib/enkf/state_map.cpp#L148-L153

I'm inclined to claim that transition from `STATE_UNDEFINED` to `STATE_LOAD_FAILURE` should be allowed, which avoids this abort and lets this line finish so that the state of the `JobQueueNode` is set properly

https://github.com/equinor/ert/blob/5770c29261ed2d069220b509fb87916a60b136cb/res/job_queue/job_queue_node.py#L203

Modifying the logic in the exit_callback [here](https://github.com/equinor/ert/blob/400611c9f2d08e61f726874dc8de7e3f2679e504/libres/lib/enkf/enkf_state.cpp#L691) to ensure states are set properly in all cases might be another option but I think it boils down to allowing the transition. Modifying the [Python calling the exit_handler](https://github.com/equinor/ert/blob/5770c29261ed2d069220b509fb87916a60b136cb/res/job_queue/job_queue_node.py#L203) does not seem to be a good idea.

**Comment by jondequinor** [comment]
If @joakim-hove has time for a review of the above comment, that would probably be very valuable. In the very least a quick history lesson on that state machine?

**Comment by joakim-hove** [comment]
I am a bit busy now - can write more later. But short:

Things might have changed in the code, so maybe allowing the transition is the right approach now. But in historical times this transition would be a very clear indication of "something broken" - and therefor a reason to fail "early". 

Without knowing details I would be a bit weary of fixing this by allowing the transition. My hip shot suggestion would be something with the queue system.

**Comment by BjarneHerland** [comment]
Another point is that the statemachine allows transition from `STATE_UNDEFINED` to `STATE_PARENT_FAILURE`. Why not allow `STATE_LOAD_FAILURE` also?

**Comment by sondreso** [comment]
No, #2756 is related to ert3 only :slightly_smiling_face:

**Comment by jondequinor** [comment]
@BjarneHerland when ERT produced `with message: state_map_iset__: illegal state transition for realisation:32740 1 -> 8` the following is the timeline of what happened on the server side:

1. `13:54:00` ERT was started
2. `13:54:38` an ensemble experiment is run
3. `13:58:17` something failed, likely `run_and_get_successful_realizations` due to some TCP/WS error
4. `13:58:17` evaluator likely shuts down here, because `run_and_get_successful_realizations` asks it to
5. `13:58:21` the job queue loop dies due to connection error in `_publish_changes`, note that the job queue _node_ loops probably remains alive
6. `14:04:13` some job queue _node_ loop sees that a forward model changed, tries to update the state

Due to 5, you're probably going to have a stale state whenever 6 happens.

I think this is a version of this: https://github.com/equinor/ert/issues/2769 but also reveals an issue in `_publish_changes` where there's literally no error handling. I've created https://github.com/equinor/ert/issues/2781

**Comment by BjarneHerland** [comment]
I am not fanatically positive to mess with established states so this sounds nice, and it aims at preventing this AFAICS
> My guess is that status as submitted has never registrered.

However, observe that if the above situation **ever** occurs (despite our solid efforts to prevent it) then the code here
https://github.com/equinor/ert/blob/f20fec48bf332f48616c3d9c3db20c27f7c2876e/res/job_queue/job_queue_node.py#L203
and possibly also here
https://github.com/equinor/ert/blob/f20fec48bf332f48616c3d9c3db20c27f7c2876e/res/job_queue/job_queue_node.py#L193
unconditionally tries to transition to illegal states, subtly failing like in this case.

Taking one step back: Can we use this state-machine more constructively than we do currently? Perhaps re-submit the job automatically (once, at least)? Log clearer and more informative information before bombing out? Or at least bomb out in such a way that the application really stops and notifies the user that something bad happened. 😄 

I unassign myself from this now as it should probably be tracked/updated in sync with the issues @jondequinor points to.

**Comment by BjarneHerland** [comment]
> So, my stab at this would be to fix #2769 and #2781 but leave the state machine alone

Since both these are closed, can we close this one also?
